File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& h) {
        int n = h.size();

        vector<int> pse(n);
        vector<int> nse(n);

        stack<int> st;
        for(int i = n - 1 ; i >= 0 ; i-- ){
            while(!st.empty() && h[st.top()] >= h[i]) st.pop();
            nse[i] = st.empty() ? n : st.top();
            st.push(i);
        }

        st = stack<int>();

        for(int i =0 ; i < n ; i++ ){
            while(!st.empty() && h[st.top()] >= h[i]) st.pop();
            pse[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }
        int maxarea = INT_MIN;
        for(int i = 0 ; i < n ; i ++){
            maxarea = max(maxarea , (nse[i] - pse[i] - 1) * h[i]);
        }
        return maxarea;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        nse = [n] * n
        pse = [-1] * n
        st = []

        for i in range(n-1 , -1 , -1):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            nse[i] = st[-1] if st else n
            st.append(i) 

        st = []
        
        for i in range(n):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            pse[i] = st[-1] if st else -1
            st.append(i)
        
        maxarea = -10**9

        for i in range(n):
            maxarea = max(maxarea , (nse[i] - pse[i] - 1)*heights[i])
        return maxarea
        